daemon/cmd/config.go

```diff
@@ -82,16 +83,25 @@ func (h *patchConfig) Handle(params PatchConfigParams) middleware.Responder {
 	changes += option.Config.Opts.ApplyValidated(om, changedOption, d)
 
 	log.WithField("count", changes).Debug("Applied changes to daemon's configuration")
-	option.Config.ConfigPatchMutex.Unlock()
 
 	if changes > 0 {
 		// Only recompile if configuration has changed.
 		log.Debug("daemon configuration has changed; recompiling base programs")
 		if err := d.Datapath().Loader().Reinitialize(d.ctx, d, d.mtuConfig.GetDeviceMTU(), d.Datapath(), d.l7Proxy); err != nil {
 			msg := fmt.Errorf("Unable to recompile base programs: %s", err)
+			// Revert configuration changes
+			if policyEnforcementChanged {
+				policy.SetPolicyEnabled(oldEnforcementValue)
+			}
+			option.Config.Opts = oldConfigOpts
+			log.Debug("finished reverting agent configuration changes")
+			option.Config.ConfigPatchMutex.Unlock()
 			return api.Error(PatchConfigFailureCode, msg)
 		}
+		option.Config.ConfigPatchMutex.Unlock()
 		d.TriggerPolicyUpdates(true, "agent configuration update")
+	} else {
+		option.Config.ConfigPatchMutex.Unlock()
```

Comment:
Same here about deferring the `RUnlock()`. Eliminates the need to have if-else branches. We know that the unlock should take place at the exit points of the function anyway.